drivers/gles2/rasterizer_canvas_gles2.cpp

```diff
@@ -45,304 +45,160 @@ static const GLenum gl_primitive[] = {
 	GL_TRIANGLE_FAN
 };
 
-RasterizerCanvasGLES2::BatchData::BatchData() {
-	reset_flush();
-	gl_vertex_buffer = 0;
-	gl_index_buffer = 0;
-	max_quads = 0;
-	vertex_buffer_size_units = 0;
-	vertex_buffer_size_bytes = 0;
-	index_buffer_size_units = 0;
-	index_buffer_size_bytes = 0;
-	use_colored_vertices = false;
-	use_light_angles = false;
-	settings_use_batching = false;
-	settings_max_join_item_commands = 0;
-	settings_colored_vertex_format_threshold = 0.0f;
-	settings_batch_buffer_num_verts = 0;
-	scissor_threshold_area = 0.0f;
-	joined_item_batch_flags = 0;
-	diagnose_frame = false;
-	next_diagnose_tick = 10000;
-	diagnose_frame_number = 9999999999; // some high number
-	join_across_z_indices = true;
-	settings_item_reordering_lookahead = 0;
-
-	settings_use_batching_original_choice = false;
-	settings_flash_batching = false;
-	settings_diagnose_frame = false;
-	settings_scissor_lights = false;
-	settings_scissor_threshold = -1.0f;
-	settings_use_single_rect_fallback = false;
-	settings_light_max_join_items = 16;
-
-	settings_uv_contract = false;
-	settings_uv_contract_amount = 0.0f;
-
-	stats_items_sorted = 0;
-	stats_light_items_joined = 0;
-}
-
-void RasterizerCanvasGLES2::RenderItemState::reset() {
-	current_clip = nullptr;
-	shader_cache = nullptr;
-	rebind_shader = true;
-	prev_use_skeleton = false;
-	last_blend_mode = -1;
-	canvas_last_material = RID();
-	item_group_z = 0;
-	item_group_light = nullptr;
-	final_modulate = Color(-1.0, -1.0, -1.0, -1.0); // just something unlikely
-
-	joined_item = nullptr;
-}
+void RasterizerCanvasGLES2::_batch_upload_buffers() {
 
-// just translate the color into something easily readable and not too verbose
-String RasterizerCanvasGLES2::BatchColor::to_string() const {
-	String sz = "{";
-	const float *data = get_data();
-	for (int c = 0; c < 4; c++) {
-		float f = data[c];
-		int val = ((f * 255.0f) + 0.5f);
-		sz += String(Variant(val)) + " ";
-	}
-	sz += "}";
-	return sz;
-}
+	// noop?
+	if (!bdata.vertices.size())
+		return;
 
-RasterizerStorageGLES2::Texture *RasterizerCanvasGLES2::_get_canvas_texture(const RID &p_texture) const {
-	if (p_texture.is_valid()) {
+	glBindBuffer(GL_ARRAY_BUFFER, bdata.gl_vertex_buffer);
 
-		RasterizerStorageGLES2::Texture *texture = storage->texture_owner.getornull(p_texture);
+	// orphan the old (for now)
+	//glBufferData(GL_ARRAY_BUFFER, 0, 0, GL_DYNAMIC_DRAW);
```

Comment:
I actually changed this just yesterday, commenting out the glBufferCall. I didn't think it was necessary, and the test versions should reveal any problems (I think I had initially put it in the first batching for safety).

Every upload uses a glBufferData call rather than glSubBufferData, so will be requesting a new buffer each time. The buffer sizes are usually quite small, so at the moment this seems more appropriate than trying to maintain a large number of larger fixed sized buffers per frame. (This is same mechanism as GLES2 batching).